src/jit/morph.cpp

```diff
@@ -3389,835 +3171,928 @@ GenTreeCall* Compiler::fgMorphArgs(GenTreeCall* call)
+        // available when we call UpdateRegArg so that we correctly update the fgArgInfo
+        // with the folded tree that represents the final optimized argument nodes.
+        //
+        if (call->gtCallLateArgs != nullptr)
+        {
 
-                    isRegArg = ((nextFltArgRegNum + structFloatRegs) <= MAX_FLOAT_REG_ARG) &&
-                               ((intArgRegNum + structIntRegs) <= MAX_REG_ARG);
-                }
-                else
-                {
-                    if (passUsingFloatRegs)
-                    {
-                        isRegArg = nextFltArgRegNum < MAX_FLOAT_REG_ARG;
-                    }
-                    else
-                    {
-                        isRegArg = intArgRegNum < MAX_REG_ARG;
-                    }
-                }
-#else  // !defined(UNIX_AMD64_ABI)
-                isRegArg = (intArgRegNum + (size - 1)) < maxRegArgs;
-#endif // !defined(UNIX_AMD64_ABI)
-#endif // _TARGET_ARM_
-            }
-            else
-            {
-                isRegArg = false;
-            }
+            call->gtCallLateArgs = fgMorphTree(call->gtCallLateArgs)->AsArgList();
+            flagsSummary |= call->gtCallLateArgs->gtFlags;
+        }
+        assert(call->fgArgInfo != nullptr);
+    }
+    call->fgArgInfo->RemorphReset();
 
-            // If there are nonstandard args (outside the calling convention) they were inserted above
-            // and noted them in a table so we can recognize them here and build their argInfo.
-            //
-            // They should not affect the placement of any other args or stack space required.
-            // Example: on AMD64 R10 and R11 are used for indirect VSD (generic interface) and cookie calls.
-            isNonStandard = nonStandardArgs.FindReg(argx, &nonStdRegNum);
-            if (isNonStandard)
-            {
-                isRegArg = (nonStdRegNum != REG_STK);
-            }
-#if defined(_TARGET_X86_)
-            else if (call->IsTailCallViaHelper())
-            {
-                // We have already (before calling fgMorphArgs()) appended the 4 special args
-                // required by the x86 tailcall helper. These args are required to go on the
-                // stack. Force them to the stack here.
-                assert(numArgs >= 4);
-                if (argIndex >= numArgs - 4)
-                {
-                    isRegArg = false;
-                }
-            }
-#endif    // defined(_TARGET_X86_)
-        } // end !reMorphing
+    // First we morph the argument subtrees ('this' pointer, arguments, etc.).
+    // During the first call to fgMorphArgs we also record the
+    // information about late arguments we have in 'fgArgInfo'.
+    // This information is used later to contruct the gtCallLateArgs */
+
+    // Process the 'this' argument value, if present.
+    argx = call->gtCallObjp;
+    if (argx)
+    {
+        fgArgTabEntry* thisArgEntry = gtArgEntryByArgNum(call, 0);
```

Comment:
I'm worried about this call, and the one below, to `gtArgEntryByArgNum()`. That function does a linear walk through the argument table looking for the matching argNum. Before, we only did this when remorphing. Now, we're doing it even for the first time processing the arguments. This seems like a bad O(n^2) case, especially in the loop below, for large argument functions. Could we void the search if we didn't sort the argTable? Or kept an unsorted table mapping index to ArgTableEntry directly?

Reply:
I am also concerned about this, though there are still a surprising number of places where this is done. In the previous version it would actually do this search **twice** in the remorphing case (it would look it up once before processing then node, then when calling the `Remorph***Arg()` method, it wouldn't bother to pass in the argEntry it already had, even though it wasn't changing it. I've added a new method, `fgArgInfo::GetArgEntry` that takes an argNum, and an optional `reMorphing` bool - if we're in the original morphing phase, we haven't yet sorted the arg table. So this eliminates the extra loops in the first round.
I also modified `fgArgInfo::EvalToTmp` to take an `fgArgTabEntry` as all the calling sites had it available already. That should leave me with fewer of these loops than before.